core/Pos/Core/Txp.hs

```diff
@@ -218,24 +224,40 @@ txaF = later $ \(TxAux tx w) ->
 instance Bi Tx => Buildable TxAux where
     build = bprint txaF
 
--- | Create valid Tx or fail.
--- Verify inputs and outputs are non empty; have enough coins.
-mkTx :: NonEmpty TxIn -> NonEmpty TxOut -> TxAttributes -> Either Text Tx
-mkTx inputs outputs attrs =
+-- | Verify inputs and outputs are non empty; have enough coins.
+checkTx
+    :: MonadError Text m
+    => Tx
+    -> m ()
+checkTx it =
     case verRes of
-        VerSuccess -> Right $ UnsafeTx inputs outputs attrs
-        failure    -> Left $ sformat verResSingleF failure
+        VerSuccess -> pure ()
+        failure    -> throwError $ sformat verResSingleF failure
   where
     verRes =
         verifyGeneric $
-        concat $ zipWith outputPredicates [0 ..] $ toList outputs
+        concat $ zipWith outputPredicates [0 ..] $ toList (_txOutputs it)
     outputPredicates (i :: Word) TxOut {..} =
-        [ ( txOutValue > mkCoin 0
+        [ ( txOutValue > Coin 0
           , sformat
                 ("output #"%int%" has non-positive value: "%coinF)
-                i txOutValue)
+                i txOutValue
+          )
+        , ( isRight (checkCoin txOutValue)
```

Comment:
I don't like that the reason why `Coin` is invalid is lost.

Reply:
Also I think adding `checkTxOut` function would be more consistent with other code.